### PR TITLE
UX: making the process of Adding and removing reactions smooth

### PR DESCRIPTION
**Pull request by Ahmedgagan**

### CODE REVIEW

File: assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6 [hunk 1]
```diff
@@ -249,14 +249,42 @@ export default createWidget("discourse-reactions-actions", {
...
-                  params.postId,
-                  params.reaction
-                ).then(resolve);
-              });
+              this.collapsePanels();
+              this.dropUserReaction();
+              this.attrs.post.reaction_users_count -= 1;
+              this.attrs.post.current_user_used_main_reaction = false;
+              this.setCurrentUserReaction();
+              setTimeout(() => {
```
Comment (jjaffeux) [inline assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6]: should user ember.run.later

Comment (jjaffeux) [reply assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6]: also you should make sure the timeout won't happen after the page has changed for example.

File: assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6 [hunk 2]
```diff
@@ -249,14 +249,42 @@ export default createWidget("discourse-reactions-actions", {
...
+                this.attrs.post.current_user_reaction.id ==
+                  this.siteSettings.discourse_reactions_like_icon
+              ) {
+                this.attrs.post.current_user_used_main_reaction = true;
+              } else {
+                this.attrs.post.current_user_used_main_reaction = false;
+              }
+
               addReaction(postContainer, params.reaction, () => {
+                setTimeout(() => {
```
Comment (jjaffeux) [inline assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6]: ember.run.later

**Comment by jjaffeux** [comment]
can you post a video of how it looks now please ?

**Comment by Ahmedgagan** [comment]
> can you post a video of how it looks now please ?


https://user-images.githubusercontent.com/35112518/103541887-1782c480-4ec2-11eb-8398-ea820340355d.mov

**Comment by jjaffeux** [comment]
something is bothering me lot, it's how height is changing and making the whole posts area moce verically on first reaction (probablu due to reactiin changing height if the parent container). We nees to fix this.

**Comment by Ahmedgagan** [comment]
In my screen I can't notice the height change😅 can you share a video the issue you are facing?

File: assets/stylesheets/common/discourse-reactions.scss [hunk 3]
```diff
@@ -255,9 +255,9 @@ html.discourse-reactions-no-select {
 
   .btn-toggle-reaction-emoji {
     align-self: center;
-    margin: 0 0.6em;
-    width: 1.1em;
-    height: 1.1em;
+    padding: 8px 8px 8px 8px;
```
Comment (jjaffeux) [inline assets/stylesheets/common/discourse-reactions.scss]: padding: 8px;

File: assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6 [hunk 4]
```diff
@@ -245,21 +245,68 @@ export default createWidget("discourse-reactions-actions", {
...
+              this.attrs.post.current_user_used_main_reaction = false;
+              this.setCurrentUserReaction();
+              later(() => {
+                dropReaction(postContainer, params.reaction, () => {
+                  return CustomReaction.toggle(
+                    params.postId,
+                    params.reaction
+                  ).then((value) => {
+                    resolve;
+                    if(value == undefined) {
```
Comment (jjaffeux) [inline assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6]: !value

Comment (jjaffeux) [inline assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6]: also please make sure prettier is correctly setup, you are missing a space here

File: assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6 [hunk 5]
```diff
@@ -245,21 +245,68 @@ export default createWidget("discourse-reactions-actions", {
...
+                this.attrs.post.current_user_used_main_reaction = false;
+              }
+
               addReaction(postContainer, params.reaction, () => {
-                CustomReaction.toggle(params.postId, params.reaction).then(
-                  resolve
-                );
+                this.collapsePanels();
+                CustomReaction.toggle(params.postId, params.reaction).then((value) => {
+                  resolve;
```
Comment (jjaffeux) [inline assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6]: that seems odd

File: assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6 [hunk 6]
```diff
@@ -245,21 +245,68 @@ export default createWidget("discourse-reactions-actions", {
...
+                  if(value == undefined) {
```
Comment (jjaffeux) [inline assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6]: js uses triple ===

Comment (jjaffeux) [reply assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6]: also !value unless you actually need to check for undefined here?

Comment (Ahmedgagan) [reply assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6]: we need to check for undefined, as !value doesn't work in our case

File: assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6 (hunk 5 above)
Comment (Ahmedgagan) [reply assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6]: when adding a reaction, the `addReaction` method creates an animation of adding reaction. so, after completion of the animation, we need to collapse the emoji picker panel.

Comment (jjaffeux) [reply assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6]: yes I know what resolve does, but that still seems odd here, will look later

File: assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6 [hunk 7]
```diff
@@ -266,7 +266,7 @@ export default createWidget("discourse-reactions-actions", {
                     params.postId,
                     params.reaction
                   ).then(value => {
-                    resolve;
+                    resolve();
```
Comment (jjaffeux) [inline assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6:269]: see :p I told you it was odd

File: app/services/discourse_reactions/reaction_manager.rb [hunk 8]
```diff
@@ -22,6 +22,7 @@ def toggle!
     private
 
     def toggle_like
+      PostAction.limit_action!(@user, @post, post_action_like_type)
```
Comment (jjaffeux) [inline app/services/discourse_reactions/reaction_manager.rb]: This will need a test

Comment (Ahmedgagan) [reply app/services/discourse_reactions/reaction_manager.rb]: I think this has no need here as its already called with `PostActionCreator`

File: assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6 (hunk 7 above)
Comment (Ahmedgagan) [reply assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6:269]: that's true😂😂

File: assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6 [hunk 9]
```diff
@@ -307,15 +362,117 @@ export default createWidget("discourse-reactions-actions", {
...
+              this.attrs.post.reaction_users_count = reaction_users_count;
+              this.scheduleRerender();
+            }
+          });
         });
       });
     });
   },
 
+  setCurrentUserReaction(reaction_id = null) {
```
Comment (jjaffeux) [inline assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6]: should be reactionId

File: assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6 [hunk 10]
```diff
@@ -307,15 +362,117 @@ export default createWidget("discourse-reactions-actions", {
...
+          return false;
+        }
+
+        return true;
+      });
+    }
+    this.scheduleRerender();
+  },
+
+  addUserReaction(reaction_id) {
```
Comment (jjaffeux) [inline assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6]: reactionId

File: assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6 [hunk 11]
```diff
@@ -245,20 +245,70 @@ export default createWidget("discourse-reactions-actions", {
...
             ) {
-              dropReaction(postContainer, params.reaction, () => {
-                return CustomReaction.toggle(
-                  params.postId,
-                  params.reaction
-                ).then(resolve);
-              });
+              this.collapsePanels();
+              this.dropUserReaction();
+              this.attrs.post.reaction_users_count -= 1;
```
Comment (jjaffeux) [inline assets/javascripts/discourse/widgets/discourse-reactions-actions.js.es6:260]: what is this -1 ?

**Review by jjaffeux — APPROVED** [review]

**Pull request merged**